### DISCUSSION TITLE
Distributed cache separate TTL option.

### DISCUSSION THREAD

**Discussion by RogerSep** [ZiggyCreatures/FusionCache#60] [Ideas]
Hi! I'd like to have a different configuration for the duration of the in memory values than the ttl in redis. Is there a way around that?

The idea is that the memory cache refreshes more often than the redis cache, so if other node updates the value in the distributed cache, the memory one will only be stalled by the duration of the in memory configuration.

Thanks! :)

**Comment by jodydonetti** [comment]
Hi @RogerSep what you are suggesting is something I've in fact already thought about, and could be done.
In reality though it would not solve the synchronization problem for real, but kind of make it less serious (which of course
can still be a valid solution in certain scenarios). What is really needed is a backplane, which I'm working on already and you can look at [here](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/issues/11).

Let me know what you think, and I can certainly do them both to cover each scenario.

Btw as I've said in the other issue I'm leaving right now for a 2 week vacation, so you'll have to wait a little bit.

**Comment by angularsen** [comment]
Similar request in #66 .

**Comment by jodydonetti** [comment]
Update: I'm currently exploring this.
It's not as straightforward as it seems for various edge cases & co but I'm working on it 😉

**Reply by angularsen** [reply]
Great! 
I'm still fairly new to distributed caching, but here are some first impressions/questions.

The concept of fail-safe mode is not super intuitive to me. I get how it works, you set an expiration time, and if you also enable fail-safe mode and set a fail-safe max duration, then it will extend the expiration time while it tries to renew the value, in the background if you wish.

I would maybe expect instead:
- Set an expiration time.
- Optionally set an eager renewal time, it will attempt to renew the value from the data source, and if successful, update both local and distributed cache - with some retries on the distributed cache.

If I am not mistaken, this would solve #66 and this discussion and maybe simplify some things.

I don't really need a different lifetime of local vs distributed cache, I just need to keep the distributed cache alive for a long time, while eagerly trying to renew local + distributed cache much earlier, and send eviction events to others when successful.

What are your thoughts on this model? @jodydonetti

**Reply by jodydonetti** [reply]
Hi @angularsen , sorry for the delay in answering but covid finally got me.

The eager renewal you described is a different approach for the same problem. In Akamai for example it is known as "cache prefreshing".

I thought about adding such a feature before, but I currently haven't found a real need for that: with FusionCache you can get the same result by setting fail-safe and a soft timeout, but without having to set an additional "eager duration" or something like that.

I see this like this, basically:
- you decide your cache duration normally, say `5 min`
- if you want to be "protected" by temporary database failures, you enable fail-safe
- if you want to be "protected" by slow database calls you enable soft/hard timeouts

and that's it: after `5 min` your data start updating, but without incurring in errors or slowness because of temporary database problems, all while keeping everything in-sync.

> I don't really need a different lifetime of local vs distributed cache, I just need to keep the distributed cache alive for a long time, while eagerly trying to renew local + distributed cache much earlier, and send eviction events to others when successful.

A quick note: to send evictions to other nodes you need to use the [backplane](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/blob/main/docs/Backplane.md).

Other than this, the 2 approaches are very similar.

In closing, as I said, I thought about adding such a feature before, but I currently haven't found a real need for that.
I mean, I get the feel of _"start updating the cache beforehand"_ but that would make sense in a cache without fail-safe and without soft/hard timeouts, where (and I think this is the important part here) **refreshing the cache beforehand would avoid database errors or delays to be reflected to the calling code** when the data actually expires, but when you think about it with fail-safe and soft/hard timeouts it becomes unnecessary because those problems go away automatically.

Let me know what you think.

**Reply by angularsen** [reply]
I don't have the setup available to verify, but I believe this is what I encountered.

What we wanted:
- Cache lifetime 24 hours (locally + Redis)
- Start eagerly renewing after 6 hours

Our config:
- Duration: 6 hours
- IsFailSafeEnabled: true
- FailSafeMaxDuration: 18 hours

Problem:
- Redis cache expired after 6 hours, instead of 24 hours.

Maybe I misunderstood or mis-configured something, but I don't think we were able to get the 24 hour expiration as expected. Do you perhaps see a flaw in my setup?

**Reply by jodydonetti** [reply]
Your config seems correct, except for a minor detail: the `FailSafeMaxDuration` set to 18 hours when it should've been 24 hours (the `FailSafeMaxDuration` is the total max duration).

So, your config should be:
- Duration: 6 hours
- IsFailSafeEnabled: true
- FailSafeMaxDuration: 24 hours

After 6 hours the first request coming in should trigger the loading of fresh data from the db, all while being able to use the old data for some more time. If you also configured a soft timeout, everything should stay fast too.

So what you described is correct, and I would not expect it to expire after just 6 hours (and the existing tests, run against a real distributed cache, confirm this).

What distributed cache are you using? Redis or something else?

If you can create a minimal repro I would be happy to take a look.

**Reply by angularsen** [reply]
@jodydonetti Sorry for the delay, but I finally got back to testing this.

I was not able to reproduce, so I must have misunderstood something during my initial implementation and testing. What generally confused me the most, is that `Duration` is not the actual cache lifetime in memcache and Redis, `FailSafeMaxDuration` is, at least when `IsFailSafeEnabled = true`.

For comparison, we have been using CacheTower and it configures cache lifetime as `timeToLive` and `staleAfter`, which at least for me was more intuitive. In particular the terms stale vs expired was helpful to me.

So with `Duration=6h, FailSafeMaxDuration=24h, IsFailSafeEnabled=true`, I successfully verified that:
- Redis cache entry does indeed have 24 hour time to live, while 6 hours is configured as metadata in the cache entry.
- When a new API/cronjob instance starts and the Redis data is stale, it does fall back to the stale data if the data source is down.


Repro (LINQPad script):
```cs
async Task Main()
{
	// Verify FusionCache behavior:
	// - Start with empty cache
	// - First GetOrSet() creates fresh value
	// - Second GetOrSet() gets memcached value
	// - Taking down source (throw exception in factory from now on)
	// - Wait 3 seconds, for cache to become stale (but not expire)
	// - Third GetOrSet() returns stale value from memcache, because factory throws
	// - Clear memcache, to simulate new application instance starting up
	// - Fourth GetOrSet() returns the same stale value, because Redis still has it cached. This mimics starting a new application, while the data source is down.
	//
	// Requires: Redis, update connection string if necessary.
	var cacheEntryOptions = new FusionCacheEntryOptions { 
		Duration = TimeSpan.FromSeconds(2),
		FailSafeMaxDuration = TimeSpan.FromSeconds(60),
		IsFailSafeEnabled = true,
	};
	var redisConnectionString = "localhost:6379";
		
	CultureInfo.CurrentCulture = CultureInfo.CurrentUICulture = CultureInfo.GetCultureInfo("en-US");

	var log = Serilog.Log.Logger = new LoggerConfiguration()
		// Use Debug level for FusionCache diagnostics.
		//.MinimumLevel.Debug()
		.MinimumLevel.Information()
		.Enrich.FromLogContext()
		.WriteTo.Console()
		.CreateLogger();
	
	var services = new ServiceCollection();
	services.AddLogging(b => b.AddSerilog(log, dispose: true));
	services.AddStackExchangeRedisCache(o => o.Configuration = redisConnectionString);
	services.AddFusionCacheSystemTextJsonSerializer();
	services.AddFusionCacheStackExchangeRedisBackplane(o => o.Configuration = redisConnectionString);
	services.AddMemoryCache();
	services.AddFusionCache(o => {
		o.DefaultEntryOptions.IsFailSafeEnabled = true;
		o.DefaultEntryOptions.FailSafeMaxDuration = TimeSpan.FromSeconds(10);
	});
	
	var sp = services.BuildServiceProvider();
	var cache = sp.GetRequiredService<IFusionCache>();
	var memCache = sp.GetRequiredService<IMemoryCache>();
	log.Information("Configuration: Duration {Duration}, FailSafeMaxDuration {FailSafeMaxDuration}\n", cacheEntryOptions.Duration, cacheEntryOptions.FailSafeMaxDuration);
	
	const string cacheKey = "myvalue";
	bool isSourceDown = false;
	int i = 0;
	int factoryCount=0;
	var sw = Stopwatch.StartNew();
	async Task<Foo> GetOrSetAsync()
	{
		var cachedValue = await cache.GetOrSetAsync(cacheKey,
			(ctx, ct) =>
			{
				log.Information($"[#{i} - {sw.Elapsed.TotalSeconds:N1} s] - Factory invoked");
				return isSourceDown
					? Task.FromException<Foo>(new Exception("Source is down"))
					: Task.FromResult(new Foo(DateTimeOffset.Now, $"[Value {factoryCount++} at {sw.Elapsed.TotalSeconds:N1} s]"));
			},
			options: cacheEntryOptions);

		return cachedValue;
	}

	void LogValue(Foo foo) => log.Information("Got {Age:N1}s old value with message {Message}", (DateTimeOffset.Now - foo.When).TotalSeconds, foo.Msg);

	log.Information("Clear any previous cached value");
	cache.Remove(cacheKey);
	log.Information("");
	
	log.Information("First GetOrSet(), should create fresh value");
	var val = await GetOrSetAsync();
	LogValue(val);
	log.Information("");
	
	log.Information("Second GetOrSet(), should get memcached value");
	val = await GetOrSetAsync();
	LogValue(val);
	log.Information("");
	
	log.Information("Taking down source");
	log.Information("Waiting 3 seconds, for cache to become stale (but not expire)");
	Thread.Sleep(3000);
	log.Information("");
	
	log.Information("Third GetOrSet() should return stale value, because factory throws");
	isSourceDown = true;
	val = await GetOrSetAsync();
	LogValue(val);
	log.Information("");
	
	log.Information("Clear memcache, to simulate new application instance starting up");
	var memcachedValue = memCache.Get(cacheKey).Dump("Memcache value");
	if (memcachedValue == null) throw new InvalidOperationException("Expected memcache to have a value before clearing.");
	memCache.Remove(cacheKey);
	memcachedValue = memCache.Get(cacheKey).Dump("Memcache value");
	if (memcachedValue != null) throw new InvalidOperationException("Memcached not removed");
	
	log.Information("Fourth GetOrSet() should return the same stale value, because Redis still has it cached. This mimics starting a new application, while the data source is down.");
	val = await GetOrSetAsync();
	LogValue(val);
	log.Information("");
}

public record Foo(DateTimeOffset When, string Msg);
```

Output:
```
[15:17:10 INF] Configuration: Duration 00:00:02, FailSafeMaxDuration 00:01:00

[15:17:10 INF] Clear any previous cached value
[15:17:10 INF] 
[15:17:10 INF] First GetOrSet(), should create fresh value
[15:17:10 INF] [#0 - 0.0 s] - Factory invoked
[15:17:10 INF] Got 0.0s old value with message [Value 0 at 0.0 s]
[15:17:10 INF] 
[15:17:10 INF] Second GetOrSet(), should get memcached value
[15:17:10 INF] Got 0.0s old value with message [Value 0 at 0.0 s]
[15:17:10 INF] 
[15:17:10 INF] Taking down source
[15:17:10 INF] Waiting 3 seconds, for cache to become stale (but not expire)
[15:17:13 INF] Third GetOrSet() should return stale value, because factory throws
[15:17:13 INF] [#0 - 3.0 s] - Factory invoked
[15:17:13 WRN] FUSION (O=d85b727e031a443f9fb81bec832b3429 K=myvalue): an error occurred while calling the factory
System.Exception: Source is down
   at ZiggyCreatures.Caching.Fusion.FusionCache.GetOrSetEntryInternalAsync[TValue](String operationId, String key, Func`3 factory, MaybeValue`1 failSafeDefaultValue, FusionCacheEntryOptions options, CancellationToken token) in /_/src/ZiggyCreatures.FusionCache/FusionCache_Async.cs:line 269
[15:17:13 WRN] FUSION (O=d85b727e031a443f9fb81bec832b3429 K=myvalue): FAIL-SAFE activated (from distributed)
[15:17:13 INF] Got 3.0s old value with message [Value 0 at 0.0 s]
[15:17:13 INF] 
[15:17:13 INF] Clear memcache, to simulate new application instance starting up
Memcache value [FFS=Y LEXP=30s]
Memcache value null
[15:17:13 INF] Fourth GetOrSet() should return the same stale value, because Redis still has it cached. This mimics starting a new application, while the data source is down.
[15:17:13 INF] [#0 - 3.0 s] - Factory invoked
[15:17:13 WRN] FUSION (O=568faef0fa23470e8efdc93146fab1b6 K=myvalue): an error occurred while calling the factory
System.Exception: Source is down
   at ZiggyCreatures.Caching.Fusion.FusionCache.GetOrSetEntryInternalAsync[TValue](String operationId, String key, Func`3 factory, MaybeValue`1 failSafeDefaultValue, FusionCacheEntryOptions options, CancellationToken token) in /_/src/ZiggyCreatures.FusionCache/FusionCache_Async.cs:line 269
[15:17:13 WRN] FUSION (O=568faef0fa23470e8efdc93146fab1b6 K=myvalue): FAIL-SAFE activated (from distributed)
[15:17:13 INF] Got 3.0s old value with message [Value 0 at 0.0 s]
[15:17:13 INF] 
```

**Reply by jodydonetti** [reply]
> What generally confused me the most, is that Duration is not the actual cache lifetime in memcache and Redis, FailSafeMaxDuration is, at least when IsFailSafeEnabled = true.

Exactly!

Here's my rationale for that: normally you would start using FusionCache like any other cache, so the cache duration is "how much time this will be valid", normal stuff.

Then you may start enabling extra features, like fail-safe: in that case duration still means "how much time this will be valid" but with the added pro that, in case something goes wrong, you can still use that data for some more time, instead of just getting an exception.
By simply enabling fail-safe nothing changes as a FusionCache user, you now just have more protection in case of problems.

From the "outside" (eg: as a user of FusionCache), duration always means the same: "for how much time this data will be valid".

From the "inside" instead, without fail-safe the duration corresponds to the actual physical expiration (now + duration) whereas with fail-safe the duration becomes more a "logical duration", and the physical expiration becomes now + fail-safe max duration, just like you described: there's no way around that, because keeping the expiration the same in both cases means FusionCache would not have anything stale to return in case something bad happens.

Given the fact that I see that more as an "internal detail", I thought it was more important to give precedence to FusionCache users, as in you should not change how you think about for how much something should be valid, you just keep doing what you were doing before enabling fail-safe.
You want to cache something for `5 min` and then have it refreshed? You do the same with or without fail-safe, nothing have to change: if you enable it, you get extra protection from transient exceptions. You also want faster response times? You enable soft/hard timeouts and you get that, without having to change durations around your codebase.

Does it make sense, at least as in how do I picture it?

Maybe I can change the docs a little bit to make that clearer. Maybe I can add an extra block in the fail-safe docs, something like "Implementation details" or something like that, where I explain more clearly that enabling fail-safe changes the physical expiration in the caches (memory and distributed).

What do you think?

> For comparison, we have been using CacheTower and it configures cache lifetime as timeToLive and staleAfter, which at least for me was more intuitive. In particular the terms stale vs expired was helpful to me.

CacheTower is a great lib!

With both CacheTower and FusionCache you can specify just a duration, and after that time passes the data will be gone and the factory will try to refresh it.

But then:
- with CacheTower you can also specify an additional duration for when the data should begin to be considered stale (and start refreshing the value, in advance)
- with FusionCache you can enable fail-safe, and be protected by problems while refreshing, and also enable soft/hard timeouts, to keep things fast

To me the big difference is that FusionCache can give you protection from factory errors and from slow factory executions, whereas with CacheTower, if the factory throw an error and/or takes too much time, you will not be protected by that.

But again I'd like to repeat: I know CacheTower and think it's a great lib!

One last thing: I thought about adding an extra feature to start refreshing cached data sooner, something like CacheTower `staleAfter` (expressed as a `TimeSpan`) or like the Akamai CDN so called "cache prefreshing" (expressed as a `%` of the specified duration): the thing is I haven't made up my mind yet if that would be the best approach, because if you want to consider something stale after `4 min` instead of `5 min`, just set the duration to `4 min` and you're good to go.

> So with Duration=6h, FailSafeMaxDuration=24h, IsFailSafeEnabled=true, I successfully verified that:
> 
> Redis cache entry does indeed have 24 hour time to live, while 6 hours is configured as metadata in the cache entry.
When a new API/cronjob instance starts and the Redis data is stale, it does fall back to the stale data if the data source is down.

Ok so it's working correctly, did I get it right?

Thanks for the discussion, I like when we can confront our ideas like this 💪

**Reply by angularsen** [reply]
Yes, everything is operating smoothly once we understood that
- Fail safe must be enabled for eager renewal to work (defaults to false).
- When fail safe is enabled, the meaning of `Duration` config value changes (logical expiration or eager renewal, instead of actual expiration).

I follow your reasoning for thinking in terms of fail-safe and fail-safe durations, it does make sense, but even after 6 months of using this library I still find it a bit unintuitive. 

Our primary use case is cache with eager renewal, so it might be why the naming and configuration setup rubs me the wrong way.

The above two bullet points is what stumbled me to begin with when first trying out the library, and in my humble opinion it adds unnecessary complexity. Tweaking the xmldoc texts for `Duration` and `FailSafeMaxDuration` helped clarify, but there is a lot of text there to explain what should be fairly simple concepts. This is an indication that maybe the configuration setup could be simplified.

We built our own setup factory methods to make it more intuitive, for us at least. This follows the naming used by CacheTower, which I immediately understood. `TimeToLive` always means the actual cache expiration and is required, while `StaleAfter` is optional and always means the logical expiration of the cache. I really do think these concepts are easier to follow, while still achieving exactly the same goals.

```cs
// With eager renewal. Fail-safe is enabled since staleAfter is configured.
var opts = new FusionCacheEntryOptions()
    .ConfigureLifeTime(timeToLive: TimeSpan.FromHours(24), 
                       staleAfter: TimeSpan.FromHours(6));

// Without eager renewal. Fail-safe is disabled because staleAfter is missing.
var opts = new FusionCacheEntryOptions()
    .ConfigureLifeTime(timeToLive: TimeSpan.FromHours(24));
```

```cs
/// <summary>
///     Extensions for configuring FusionCache.
/// </summary>
public static class FusionCacheEntryOptionsExtensions
{
    /// <summary>
    ///     Configures FusionCache entry lifetimes based on <see cref="timeToLive" /> and optionally <see cref="staleAfter" /> for eager renewal.
    /// </summary>
    /// <param name="options">The options to change.</param>
    /// <param name="timeToLive">How long it should live in the cache.</param>
    /// <param name="staleAfter">How long until it is considered stale and should be renewed. Must be shorter than <see cref="timeToLive" />.</param>
    /// <remarks>
    ///     FusionCache is not too intuitive to configure for eager renewal, since it has the concept of fail-safe mode, which affects how
    ///     <see cref="FusionCacheEntryOptions.Duration" /> works.
    /// </remarks>
    /// <returns>FusionCache entry options.</returns>
    public static FusionCacheEntryOptions ConfigureLifeTime(this FusionCacheEntryOptions options, TimeSpan timeToLive, TimeSpan? staleAfter = null)
    {
        if (staleAfter is { } staleAfterNotNull)
        {
            // When fail-safe mode is enabled, Duration is changed to mean when the cache is considered stale, while FailSafeMaxDuration is when it expires.
            options.IsFailSafeEnabled = true;
            options.FailSafeMaxDuration = timeToLive;
            options.Duration = staleAfterNotNull;
        }
        else
        {
            // Disable fail-safe mode if cache should be considered fresh until it expires, so FailSafeMaxDuration's default value does not have an effect.
            options.IsFailSafeEnabled = false;
            options.Duration = timeToLive;
        }

        return options;
    }

    /// <summary>
    ///     Configures FusionCache entry lifetimes based on <see cref="CacheSetup" />.
    /// </summary>
    /// <param name="options">The options to change.</param>
    /// <param name="cacheSetup">The cache setup.</param>
    /// <remarks>
    ///     FusionCache is not too intuitive to configure for eager renewal, since it has the concept of fail-safe mode, which affects how
    ///     <see cref="FusionCacheEntryOptions.Duration" /> works.
    /// </remarks>
    /// <returns>FusionCache entry options.</returns>
    public static FusionCacheEntryOptions ConfigureLifeTime(this FusionCacheEntryOptions options, CacheSetup cacheSetup)
    {
        return ConfigureLifeTime(options, cacheSetup.TimeToLive, cacheSetup.StaleAfter);
    }
}
```

**Reply by jodydonetti** [reply]
Hi @angularsen and thank you very much for sharing your thoughts!
By reflecting on them and other people's thoughts, you all got me thinking lately and I came up with [an idea](https://github.com/ZiggyCreatures/FusionCache/issues/92).

If you have some time, please read it and let me know what you think.

Thanks!

**Comment by jodydonetti** [comment]
Hi @RogerSep , I've finally added support for a custom duration for the distributed cache (`TimeSpan? DistributedCacheDuration`) which, if specified, will override the "main" duration.

This feature is included in the newly release [v0.12.0](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/releases/tag/v0.12.0).

Hope this helps.
